### PR TITLE
docs: crud vector kv indexer

### PR DESCRIPTION
**Pull request by florian-hoenicke**

### CODE REVIEW

File: jina/drivers/index.py [hunk 1]
```diff
@@ -17,7 +17,9 @@ def __init__(self, executor: str = None, method: str = 'add', *args, **kwargs):
 
 
 class VectorIndexDriver(BaseIndexDriver):
-    """Extract chunk-level embeddings and add it to the executor
+    """Extracts embeddings and ids from the documents or chunks and forwards them to the executor.
```
Comment (JoanFM) [inline jina/drivers/index.py:20]: we do not talk about chunks anymore except in specific cases.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/cache.py [hunk 2]
```diff
@@ -95,7 +99,7 @@ def query(self, data, *args, **kwargs) -> Optional[bool]:
         return status
 
     def update(self, keys: Iterator['UniqueId'], values: Iterator[any], *args, **kwargs):
-        """
+        """Update cached documents.
         :param keys: list of Document.id
         :param values: list of either `id` or `content_hash` of :class:`Document"""
```
Comment (nan-wang) [inline jina/executors/indexers/cache.py:104]:
```suggestion
        :param values: list of either `id` or `content_hash` of :class:`Document
        """
```

File: jina/executors/indexers/__init__.py [hunk 3]
```diff
@@ -53,6 +53,18 @@ def __init__(self,
         self._size = 0
 
     def add(self, *args, **kwargs):
+        """ Add documents to the index.
```
Comment (nan-wang) [inline jina/executors/indexers/__init__.py:56]:
```suggestion
        """Add documents to the index.
```

File: jina/executors/indexers/__init__.py [hunk 4]
```diff
@@ -53,6 +53,18 @@ def __init__(self,
...
+        """
+        raise NotImplementedError
+
+    def update(self, *args, **kwargs):
+        """ Update documents on the index.
```
Comment (nan-wang) [inline jina/executors/indexers/__init__.py:61]:
```suggestion
        """Update documents on the index.
```

File: jina/executors/indexers/__init__.py [hunk 5]
```diff
@@ -53,6 +53,18 @@ def __init__(self,
...
+        """
+        raise NotImplementedError
+
+    def delete(self, *args, **kwargs):
+        """ Delete documents from the index.
```
Comment (nan-wang) [inline jina/executors/indexers/__init__.py:66]:
```suggestion
        """Delete documents from the index.
```

File: jina/executors/indexers/__init__.py [hunk 6]
```diff
@@ -62,6 +74,8 @@ def post_init(self):
         self.is_handler_loaded = False
 
     def query(self, *args, **kwargs):
+        """ Query documents from the index.
```
Comment (nan-wang) [inline jina/executors/indexers/__init__.py:77]:
```suggestion
        """Query documents from the index.
```

File: jina/executors/indexers/vector.py [hunk 7]
```diff
@@ -206,9 +220,10 @@ def raw_ndarray(self) -> Optional['np.ndarray']:
 
     def query_by_id(self, ids: Union[List[int], 'np.ndarray'], *args, **kwargs) -> Optional['np.ndarray']:
         """
-        Search the index by the external key (passed during `.add(`)
+        Search the index by the external key (passed during `.add(`).
```
Comment (nan-wang) [inline jina/executors/indexers/vector.py:223]: put in the same line to stay consistent.

File: jina/drivers/index.py [hunk 8]
```diff
@@ -32,7 +34,7 @@ def _apply_all(self, docs: 'DocumentSet', *args, **kwargs) -> None:
 
 
 class KVIndexDriver(BaseIndexDriver):
-    """Serialize the documents/chunks in the request to key-value JSON pairs and write it using the executor
+    """Forwards pairs of serialized documents or chunks and ids to the executor.
```
Comment (nan-wang) [inline jina/drivers/index.py:37]: same issue as above. no more `chunks`

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/__init__.py [hunk 9]
```diff
@@ -212,25 +212,23 @@ class BaseVectorIndexer(BaseIndexer):
...
+        :return: subset of indexed vectors
         """
         raise NotImplementedError
 
     def add(self, keys: 'np.ndarray', vectors: 'np.ndarray', *args, **kwargs):
-        """Add new chunks and their vector representations
+        """Add vectors to the index.
 
-        :param keys: ``chunk_id`` in 1D-ndarray, shape B x 1
+        :param keys: document or chunk ids` as 1D-ndarray
```
Comment (JoanFM) [inline jina/executors/indexers/__init__.py:225]: no chunk here

File: jina/executors/indexers/__init__.py [hunk 10]
```diff
@@ -240,9 +238,18 @@ def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.nd
         raise NotImplementedError
 
     def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        """Update vectors on the index.
+
+        :param keys: document or chunk ids` as 1D-ndarray
```
Comment (JoanFM) [inline jina/executors/indexers/__init__.py:243]: same here

File: jina/executors/indexers/__init__.py [hunk 11]
```diff
@@ -240,9 +238,18 @@ def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.nd
...
+        :param values: vector representations in B x D
+        """
         raise NotImplementedError
 
     def delete(self, keys: Iterator[int], *args, **kwargs):
+        """Delete vectors from the index.
+
+        :param keys: document or chunk ids` as 1D-ndarray
```
Comment (JoanFM) [inline jina/executors/indexers/__init__.py:251]: and here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/drivers/__init__.py [hunk 12]
```diff
@@ -290,8 +290,8 @@ def __call__(self, *args, **kwargs):
         self._traverse_apply(self.docs, *args, **kwargs)
 
     def _traverse_apply(self, docs: 'DocumentSet', *args, **kwargs) -> None:
-        for path in self._traversal_paths:
-            if path[0] == 'r':
+        for path in self._traversal_paths:    # self._traversal_paths = ['cm']
```
Comment (JoanFM) [inline jina/drivers/__init__.py:293]: remember to remove this

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/cache.py [hunk 13]
```diff
@@ -67,6 +67,10 @@ def __init__(self, index_filename: str = None, *args, **kwargs):
             raise ValueError(f"Field '{self.field}' not in supported list of {self.supported_fields}")
 
     def add(self, doc_id: 'UniqueId', *args, **kwargs):
+        """Add a document id to the cache.
```
Comment (cristianmtr) [inline jina/executors/indexers/cache.py:70]:
```suggestion
        """Add a document to the cache (depending on field)
```

File: jina/executors/indexers/cache.py [hunk 14]
```diff
@@ -67,6 +67,10 @@ def __init__(self, index_filename: str = None, *args, **kwargs):
...
+
+        :param doc_id: Document id to be added.
```
Comment (cristianmtr) [inline jina/executors/indexers/cache.py:72]:
```suggestion
        :param doc_id: id of Document to be added.
```

File: jina/executors/indexers/keyvalue.py [hunk 15]
```diff
@@ -55,6 +55,11 @@ def __init__(self, *args, **kwargs):
         self._page_size = mmap.ALLOCATIONGRANULARITY
 
     def add(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        """ Add the serialized chunks or documents to the index via document ids.
```
Comment (cristianmtr) [inline jina/executors/indexers/keyvalue.py:58]: Remove mention of chunks here too?

File: jina/drivers/search.py [hunk 16]
```diff
@@ -31,24 +31,25 @@ def __init__(
...
+        This driver runs a query for each document.
+        This may not be very efficient, as the total number of queries grows cubic with the number of documents, chunks
+        per document and top-k.
 
-             - ``level=chunk``: D x C x K
-             - ``level=doc``: D x K
-             - ``level=all``: D x C x K
+             - ``granularity=0``: D x K
+             - ``granularity=1``: D x C x K
+             - ``granularity=1``: D x C x C x K
```
Comment (JoanFM) [inline jina/drivers/search.py:43]: this is `granularity 0 and 1`.

File: jina/drivers/search.py [hunk 17]
```diff
@@ -31,24 +31,25 @@ def __init__(
...
-        This may not be very efficient, as the total number of queries depends on ``level``
+        This driver runs a query for each document.
+        This may not be very efficient, as the total number of queries grows cubic with the number of documents, chunks
+        per document and top-k.
 
-             - ``level=chunk``: D x C x K
-             - ``level=doc``: D x K
-             - ``level=all``: D x C x K
+             - ``granularity=0``: D x K
+             - ``granularity=1``: D x C x K
```
Comment (JoanFM) [inline jina/drivers/search.py:42]: I would reflect it with `traversal_paths` arguments and not granularity. 

If 

`traversal_paths = ['m'] => D x K`
`traversal_paths = ['r'] => D`
`traversal_paths = ['cm'] => D x C x K`
`traversal_paths = ['m', 'cm'] => D x K + D x C x K`

Is this what we want to express?

File: jina/executors/indexers/__init__.py [hunk 18]
```diff
@@ -12,7 +12,7 @@
 
 
 class BaseIndexer(BaseExecutor):
-    """ base class for storing and searching any kind of data structure
+    """base class for storing and searching any kind of data structure
```
Comment (JoanFM) [inline jina/executors/indexers/__init__.py:15]: Capital letter at the beginning?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/drivers/search.py (hunk 17 above)
Comment (florian-hoenicke) [reply jina/drivers/search.py:42]: really awesome illustration. Thanks I will take this. This is much more understandable for the user

File: jina/executors/indexers/__init__.py (hunk 18 above)
Comment (florian-hoenicke) [reply jina/executors/indexers/__init__.py:15]: check

File: jina/drivers/cache.py [hunk 19]
```diff
@@ -42,7 +41,7 @@ def _apply_all(self, docs: 'DocumentSet', *args, **kwargs) -> None:
                     self.on_hit(d, result)
 
     def on_miss(self, doc: 'Document', data) -> None:
-        """Function to call when doc is missing, the default behavior is add to cache when miss
+        """Function to call when document is missing, the default behavior is add to cache when miss.
```
Comment (Yongxuanzhang) [inline jina/drivers/cache.py:44]: 'the default behavior is add to cache when miss. '
does it mean 'the default behavior is to add the document to cache when miss. '?

File: jina/drivers/search.py [hunk 20]
```diff
@@ -31,24 +31,26 @@ def __init__(
 
 
 class KVSearchDriver(BaseSearchDriver):
-    """Fill in the doc/chunk-level top-k results using the :class:`jina.executors.indexers.meta.BinaryPbIndexer`
+    """Fill in the results using the :class:`jina.executors.indexers.meta.BinaryPbIndexer`
```
Comment (Yongxuanzhang) [inline jina/drivers/search.py:34]: Probably we could all use first-person expressions(like 'fill') or third person(like 'fills')? Now some docstrings use first-person but some use third-person

Comment (nan-wang) [reply jina/drivers/search.py:34]: Let's stick to the first-person expression

**Review by nan-wang — APPROVED** [review]

**Pull request merged**